src/components/cuda_pw/cupti_profiler.c

```diff
@@ -0,0 +1,1112 @@
+    NVPW_CounterDataBuilder_GetCounterDataPrefixPtr = DLSYM_AND_CHECK(dl4, "NVPW_CounterDataBuilder_GetCounterDataPrefix");
+    NVPW_CounterData_GetNumRangesPtr = DLSYM_AND_CHECK(dl4, "NVPW_CounterData_GetNumRanges");
+    NVPW_Profiler_CounterData_GetRangeDescriptionsPtr = DLSYM_AND_CHECK(dl4, "NVPW_Profiler_CounterData_GetRangeDescriptions");
+    NVPW_MetricsContext_SetCounterDataPtr = DLSYM_AND_CHECK(dl4, "NVPW_MetricsContext_SetCounterData");
+    NVPW_MetricsContext_EvaluateToGpuValuesPtr = DLSYM_AND_CHECK(dl4, "NVPW_MetricsContext_EvaluateToGpuValues");
+    NVPW_RawMetricsConfig_GetNumPassesPtr = DLSYM_AND_CHECK(dl4, "NVPW_RawMetricsConfig_GetNumPasses");
+    NVPW_RawMetricsConfig_SetCounterAvailabilityPtr = DLSYM_AND_CHECK(dl4, "NVPW_RawMetricsConfig_SetCounterAvailability");
+    NVPW_RawMetricsConfig_IsAddMetricsPossiblePtr = DLSYM_AND_CHECK(dl4, "NVPW_RawMetricsConfig_IsAddMetricsPossible");
+    NVPW_MetricsContext_GetCounterNames_BeginPtr = DLSYM_AND_CHECK(dl4, "NVPW_MetricsContext_GetCounterNames_Begin");
+    NVPW_MetricsContext_GetCounterNames_EndPtr = DLSYM_AND_CHECK(dl4, "NVPW_MetricsContext_GetCounterNames_End");
+
+fn_exit:
+    return papiErr;
+fn_fail:
+    papiErr = PAPI_ENOSUPP;
+    goto fn_exit;
+
+}
+
+static int initialize_cupti_profiler_API(void)
+{
+    COMPDBG("Entering.\n");
+    CUpti_Profiler_Initialize_Params profilerInitializeParams = {CUpti_Profiler_Initialize_Params_STRUCT_SIZE, NULL };
+    if (cuptiProfilerInitializePtr(&profilerInitializeParams) == CUPTI_SUCCESS)
+        return PAPI_OK;
+    else
+        return PAPI_ESYS;  // or something else?
+}
+
+static int initialize_perfworks_API(void)
+{
+    COMPDBG("Entering.\n");
+    NVPW_InitializeHost_Params perfInitHostParams = { NVPW_InitializeHost_Params_STRUCT_SIZE, NULL };
+    if (NVPW_InitializeHostPtr(&perfInitHostParams) == NVPA_STATUS_SUCCESS)
+        return PAPI_OK;
+    else
+        return PAPI_ESYS;
+    // NVPW_InitializeTarget_Params perfInitTargetParams = {NVPW_InitializeTarget_Params_STRUCT_SIZE };
+    // NVPW_CALL(NVPW_InitializeTarget(&perfInitTargetParams));
+    return PAPI_OK;
+}
+
+static int get_chip_name(int dev_num, char* chipName)
+{
+    CUpti_Device_GetChipName_Params getChipName = {
+        .structSize = CUpti_Device_GetChipName_Params_STRUCT_SIZE,
+        .pPriv = NULL,
+        .deviceIndex = 0
+    };
+    getChipName.deviceIndex = dev_num;
+    CUPTI_CALL( cuptiDeviceGetChipNamePtr(&getChipName), return PAPI_ENOSUPP );
+    strcpy(chipName, getChipName.pChipName);
+    return PAPI_OK;
+}
+
+struct byte_array_s {
+    uint8_t *data;
+    int size;
+};
+
+struct cupti_gpu_control_s {
+    int gpu_id;
+    event_list_t event_names;
+    int rmr_count;
+    NVPA_RawMetricRequest *rmr;
+    struct byte_array_s counterDataImagePrefix,
+                        configImage,
+                        counterDataImage,
+                        counterDataScratchBuffer,
+                        counterAvailabilityImage;
```

Comment:
Don't write things like this. Add another line for each definition.

Reply:
done.